uspace/lib/cpp/include/__bits/thread/shared_state.hpp

```diff
@@ -0,0 +1,409 @@
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef LIBCPP_BITS_THREAD_SHARED_STATE
+#define LIBCPP_BITS_THREAD_SHARED_STATE
+
+/**
+ * 30.6.4, shared state:
+ */
+
+#include <__bits/exception.hpp>
+#include <__bits/functional/function.hpp>
+#include <__bits/functional/invoke.hpp>
+#include <__bits/refcount_obj.hpp>
+#include <__bits/thread/future_common.hpp>
+#include <__bits/thread/threading.hpp>
+#include <cerrno>
+#include <thread>
+#include <tuple>
+
+namespace std::aux
+{
+    class shared_state_base: public aux::refcount_obj
+    {
+        public:
+            shared_state_base()
+                : mutex_{}, condvar_{}, value_set_{false},
+                  exception_{}, has_exception_{false}
+            {
+                threading::mutex::init(mutex_);
+                threading::condvar::init(condvar_);
+            }
+
+            void destroy() override
+            {
+                /**
+                 * Note: No need to act in this case, async shared
+                 *       state is the object that needs to sometimes
+                 *       invoke its payload.
+                 */
+            }
+
+            void mark_set(bool set = true) noexcept
+            {
+                value_set_ = set;
+            }
+
+            bool is_set() const noexcept
+            {
+                return value_set_;
+            }
+
+            void set_exception(exception_ptr ptr, bool set = true)
+            {
+                exception_ = ptr;
+                has_exception_ = set;
+            }
+
+            bool has_exception() const noexcept
+            {
+                return has_exception_;
+            }
+
+            void throw_stored_exception() const
+            {
+                if (has_exception_)
+                    rethrow_exception(exception_);
+            }
+
+            /**
+             * TODO: This member function is supposed to be marked
+             *       as 'const'. In such a case, however, we cannot
+             *       use the underlying fibril API because these
+             *       references get converted to pointers and the API
+             *       does not accept e.g. 'const fibril_condvar_t*'.
+             *
+             *       The same applies to the wait_for and wait_until
+             *       functions.
+             */
```

Comment:
Why not use `const_cast`? This is what would be called "interior mutability" in Rust parlance.

Reply:
Good catch, forgot about that, will add it.